<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px"> 

## Lab: SQL Practice

_Author: Matt Brems_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [1]:
# Import pandas and create_engine from `sqlalchemy`

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data
engine = \
create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [ ]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(*)
FROM stores
"""

pd.read_sql_query(sql, engine)

**Prompt 1 Answer**: 1973 rows

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [ ]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT store,name
FROM stores
ORDER BY 2 ASC
"""

pd.read_sql_query(sql, engine)

**Prompt 2 Answer**: 3061 Cub Foods / Sioux City

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [127]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price::MONEY,bottle_qty,vendor
FROM sales
WHERE category_name = 'IMPORTED VODKA'
ORDER BY 3 ASC

LIMIT 3
"""

pd.read_sql_query(sql, engine)

btl_price  bottle_qty                     vendor
0     $8.99          12  Adamba Imports Int'l Inc.
1    $28.50           3      A Hardy / U.S.A. Ltd.
2    $28.50           6      A Hardy / U.S.A. Ltd.

**Prompt 3 Answer**: 

- $8.99	12	Adamba Imports Int'l Inc.

- $28.50	3	A Hardy / U.S.A. Ltd.

- $28.50	6	A Hardy / U.S.A. Ltd.


### Prompt 4

What is the total amount of liquor sold in February 2015?

In [128]:
# Use a SQL query to find the answer to the above prompt.
# SELECT date,SUM(bottle_qty)

sql = """
SELECT SUM(bottle_qty)
FROM(SELECT date,SUM(bottle_qty) as bottle_qty
FROM sales
WHERE date between '2015-02-01' and '2015-02-28'
GROUP BY date) as temp
WHERE DATE_PART('year',date) = 2015 

"""
pd.read_sql_query(sql, engine)

sum
0  1591962.0

In [131]:
sql = """
SELECT SUM(bottle_qty) as bottle_qty, SUM(total) as total_sales
FROM sales
WHERE date between '2015-02-01' and '2015-02-28'
 

"""
pd.read_sql_query(sql, engine)

bottle_qty  total_sales
0     1591962  21295350.55

**Prompt 4 Answer**: 

- bottle_qty	 1591962
- total_sales  21295350.55

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [133]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT vendor,COUNT(*) as num_transaction ,COUNT(bottle_qty) as bottle_solds,SUM(total) as total_revenue
FROM sales
WHERE category_name = 'IMPORTED VODKA'
GROUP BY vendor
ORDER BY 2 DESC

LIMIT 5
"""

pd.read_sql_query(sql, engine)

vendor  num_transaction  bottle_solds  \
0  Pernod Ricard USA/Austin Nichols            42338         42338   
1               Bacardi U.S.A. Inc.            26441         26441   
2                   Diageo Americas            15725         15725   
3   Constellation Wine Company Inc.            12565         12565   
4                   Jim Beam Brands             5850          5850   

   total_revenue  
0     7880365.88  
1     5771242.07  
2     3346425.42  
3     1895005.51  
4     1807097.23

**Prompt 5 Answer**: 
 - Pernod Ricard USA/Austin Nichols	42338	42338	7880365.88

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [135]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT (btl_price-state_btl_cost) as Markup
FROM sales

LIMIT 5
"""

pd.read_sql_query(sql, engine)

markup
0  $5.75
1  $5.75
2  $5.75
3  $8.58
4  $2.50

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [137]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT store, AVG(Markup/state_btl_cost) as AVG_Percentage_Markup
FROM(SELECT (btl_price-state_btl_cost) as Markup, state_btl_cost,store
FROM sales) as temp
GROUP BY store
ORDER BY 2 DESC

LIMIT 5
"""

pd.read_sql_query(sql, engine)

store  avg_percentage_markup
0   4013               0.525471
1   4024               0.521499
2   4266               0.518903
3   3833               0.515094
4   3660               0.514591

**Prompt 7 Answer**: store number 4013 with 0.525471 AVG_Percentage_Markup

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [138]:
# Use a SQL query to find the answer to the above prompt.



sql = """
SELECT store,name,store_address,max,min,avg
FROM(
SELECT s.store,name,store_address, MIN((btl_price-state_btl_cost)/state_btl_cost) OVER (partition by s.store) as min,\
MAX((btl_price-state_btl_cost)/state_btl_cost) OVER (partition by s.store) as max, AVG((btl_price-state_btl_cost)/state_btl_cost) OVER (partition by s.store) as avg
FROM sales as s 
JOIN stores as st
ON s.store = st.store
) as temp
GROUP BY store,name,store_address,max,min,avg
ORDER BY 6 DESC
LIMIT 5


"""


pd.read_sql_query(sql, engine)

store                        name  \
0   4013      Wal-Mart 0841 / Tipton   
1   4024  Wal-Mart 1546 / Iowa Falls   
2   4266  Wal-Mart 1683 / Shenandoah   
3   3833    Wal-Mart 3394 / Atlantic   
4   3660   Wal-Mart 2935 / Knoxville   

                                       store_address       max       min  \
0  1126 Highway 38 North\nTipton, IA 527720000\n(...  6.708763  0.498571   
1  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...  6.708763  0.498462   
2  705 S Fremont\nShenandoah, IA 516010000\n(40.7...  6.708763  0.498462   
3  1905 East 7th St\nAtlantic, IA 500220000\n(41....  6.708763  0.498462   
4  814 W Bell Ave\nKnoxville, IA 501380000\n(41.3...  6.708763  0.498462   

        avg  
0  0.525471  
1  0.521499  
2  0.518903  
3  0.515094  
4  0.514591

**Prompt 8 Answer**:
1. Wal-Mart 0841 / Tipton
2. Wal-Mart 1546 / Iowa Falls
3. Wal-Mart 1683 / Shenandoah	
4. Wal-Mart 3394 / Atlantic
5. Wal-Mart 2935 / Knoxville


### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery).

In [27]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT c.county,SUM(bottle_qty)/population as bot_per_capita, ROUND(AVG(bottle_size),1) as avg_bottle_sz
FROM counties as c
JOIN sales as s
ON c.county = s.county
JOIN products as p
ON p.item_no = s.item

GROUP BY c.county
ORDER BY 2 DESC
LIMIT 9


"""

pd.read_sql_query(sql, engine)

county  bot_per_capita  avg_bottle_sz
0      Dickinson              22          974.1
1           Polk              15          892.6
2     Black Hawk              15          862.9
3    Cerro Gordo              14          986.4
4          Scott              13          871.6
5        Johnson              13          895.4
6           Linn              12          886.3
7  Pottawattamie              11          893.7
8        Kossuth              11         1041.0

**Prompt 9 Answer**: 
1. Dickinson	
2. Polk	
3. Black Hawk
4. Cerro Gordo	
5. Scott	
6. Johnson	
7. Linn	
8. Pottawattamie	
9. Kossuth	 -- largest bottle size



### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [47]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(*) 
FROM(SELECT btl_price,bottle_qty,total::MONEY,(btl_price * bottle_qty)::MONEY as check_total
FROM sales) as temp
WHERE check_total != total



"""
pd.read_sql_query(sql, engine)

count
0      0

**Prompt 10 Answer**: 0

# EXTRA QUESTIONS

In [48]:
# How many total products are in the Products table?
sql = """
SELECT COUNT(DISTINCT item_no) 
FROM products

"""
pd.read_sql_query(sql, engine)

count
0   9977

In [60]:
# Who are the top most diverse vendors (i.e. they have the highest number of distinct products)? 
# How many different products do they have?

sql = """
SELECT vendor_name, count_item,(count_item - lead_count) as DIFF
FROM(SELECT vendor_name, count_item, COALESCE(LEAD(count_item) OVER (),0) as lead_count
FROM(SELECT vendor_name,COUNT(DISTINCT item_no) as count_item
FROM products
GROUP BY vendor_name
ORDER BY 2 DESC

LIMIT 5) as temp) as temp

"""
pd.read_sql_query(sql, engine)

vendor_name  count_item  diff
0                   Jim Beam Brands         925    18
1                   Diageo Americas         907   308
2  Pernod Ricard Usa/austin Nichols         599    20
3         Yahara Bay Distillers Inc         579   191
4     Heaven Hill Distilleries Inc.         388   388

In [66]:
# Which products sell the best by total number of unit sales? 
# (Hint: you will need to find a way of doing a cross-join between the sales and product tables.)

sql = """
SELECT item_no,SUM(item) as total_item
FROM products as p 
JOIN sales as s
ON s.item = p.item_no
GROUP BY item_no
ORDER BY 2 DESC
LIMIT 5
"""
pd.read_sql_query(sql, engine)

item_no  total_item
0    36308  1129360340
1    64866  1104213918
2    89196   887500200
3    35918   831429864
4    86886   786057642

In [68]:
# Which products sell the best by total dollar value of sales?
sql = """
SELECT item_no,SUM(total) as total_values
FROM products as p 
JOIN sales as s
ON s.item = p.item_no
GROUP BY item_no
ORDER BY 2 DESC
LIMIT 5
"""
pd.read_sql_query(sql, engine)

item_no  total_values
0    11788   12863376.81
1    43337    8523370.55
2    26827    6668863.94
3    11297    6231299.41
4    36308    5380753.20

In [70]:
# What are the top 10 categories of liquor sold based on the total amount of sales revenue?
sql = """
SELECT s.category_name,SUM(total) as total_values
FROM products as p 
JOIN sales as s
ON s.item = p.item_no
GROUP BY s.category_name
ORDER BY 2 DESC
LIMIT 10

"""
pd.read_sql_query(sql, engine)

category_name  total_values
0                 CANADIAN WHISKIES   48053061.91
1                    80 PROOF VODKA   48045034.70
2                        SPICED RUM   31600618.50
3                    IMPORTED VODKA   23879524.63
4                           TEQUILA   21409700.64
5         STRAIGHT BOURBON WHISKIES   20904099.16
6                   WHISKEY LIQUEUR   19338225.94
7                TENNESSEE WHISKIES   17602242.30
8  PUERTO RICO & VIRGIN ISLANDS RUM   12729072.76
9                  BLENDED WHISKIES   12018198.55

In [ ]:
sql = """
SELECT SUM(bottle_qty)
FROM sales as s
JOIN counties as c
ON c.county = s.county
WHERE category_name LIKE '%vodka%' AND DATE_PART('month',date) = 02 AND DATE_PART('year',date) = 2014
GROUP BY s.county 
"""
pd.read_sql_query(sql,engine)

In [196]:
# Which county sold the most amount of vodka during February 2014?


sql = """
SELECT s.county,SUM(bottle_qty)
FROM sales as s
JOIN counties as c
ON c.county = s.county
WHERE category_name ILIKE '%%vodka%%' AND DATE_PART('month',date) = 02 AND DATE_PART('year',date) = 2014
GROUP BY s.county 

ORDER BY 2 DESC
LIMIT 2
"""


pd.read_sql_query(sql,engine)

county     sum
0   Polk  202012
1   Linn   71436

In [197]:
# Which counties were in the top 10 counties for vodka sales in any month in 2014?

sql = """
SELECT s.county,SUM(bottle_qty)
FROM sales as s
JOIN counties as c
ON c.county = s.county
WHERE category_name ILIKE '%%vodka%%' AND DATE_PART('year',date) = 2014
GROUP BY s.county 

ORDER BY 2 DESC
LIMIT 10
"""


pd.read_sql_query(sql,engine)

county      sum
0           Polk  1961542
1           Linn   733915
2          Scott   681223
3     Black Hawk   501407
4        Johnson   499055
5  Pottawattamie   283184
6        Dubuque   241364
7       Woodbury   233802
8          Story   232601
9    Cerro Gordo   172833

In [239]:
# Create a report that shows how many times a county appeared in the “top 10 counties for vodka sales in a month” 
# list over the course of 2014.

sql = """
SELECT county,COUNT (county)
FROM(SELECT county,month,total_bottle_month,row_number
FROM(SELECT county,month,total_bottle_month, ROW_NUMBER()OVER(partition by month)
FROM(SELECT s.county,DATE_PART('month',date) as month,SUM(bottle_qty) as total_bottle_month
FROM sales as s
JOIN counties as c
ON c.county = s.county
WHERE category_name ILIKE '%%vodka%%' and DATE_PART('year',date) = 2014 
GROUP BY s.county,DATE_PART('month',date)
ORDER BY 2 ASC, 3 DESC) as temp) as temp2
WHERE row_number <= 10) as temp3
GROUP BY temp3.county
ORDER BY 2 DESC

LIMIT 50

"""


pd.read_sql_query(sql,engine)





county  count
0        Woodbury     12
1     Cerro Gordo     12
2   Pottawattamie     12
3           Scott     12
4      Black Hawk     12
5         Dubuque     12
6         Johnson     12
7            Linn     12
8            Polk     12
9           Story     10
10         Dallas      1
11     Des Moines      1

In [296]:
# What is the trend of sales by month?
# Break up variables such as bottle_price or liter_size into categories 
# (for example: cheap, medium, or expensive).

sql = """
SELECT month, CASE WHEN total_bottle_qty <= 1000000 THEN 'low'
WHEN total_bottle_qty <= 2500000 THEN 'medium'
WHEN total_bottle_qty <= 4000000 THEN 'large'
ELSE 'very large' END AS amount_sale

FROM(SELECT date_part as month,SUM(sum) as total_bottle_qty
FROM(SELECT DATE_PART('month',date),SUM(bottle_qty)
FROM sales
GROUP BY date) as temp
GROUP BY date_part) as temp_2
ORDER BY 1 ASC ,2 ASC


LIMIT 50
"""


pd.read_sql_query(sql,engine)



month amount_sale
0     1.0  very large
1     2.0  very large
2     3.0       large
3     4.0       large
4     5.0       large
5     6.0      medium
6     7.0      medium
7     8.0      medium
8     9.0      medium
9    10.0      medium
10   11.0      medium
11   12.0         low

In [308]:
# Which stores sell one of the top five most expensive bottles of alcohol?



sql = """
SELECT vendor, btl_price
FROM sales
WHERE category_name ILIKE '%%alcohol%%'
ORDER BY 2 DESC
LIMIT 5
"""

pd.read_sql_query(sql,engine)

vendor btl_price
0  Luxco-St Louis    $35.01
1  Luxco-St Louis    $28.01
2  Luxco-St Louis    $28.01
3  Luxco-St Louis    $28.01
4  Luxco-St Louis    $28.01

In [321]:
# How many stores have more than $2,000,000 in total sales?



sql = """
SELECT COUNT(*)
FROM(SELECT store,SUM(total)
FROM sales
GROUP BY store
HAVING SUM(total) > 2000000) as temp

"""

pd.read_sql_query(sql,engine)

count
0     24

In [329]:
# How many stores have an average bottle price greater than $20?

sql = """
SELECT COUNT(*)
FROM(SELECT store, avg(btl_price::numeric)
FROM sales
GROUP BY store
HAVING avg(btl_price::numeric) > 20) as temp

"""

pd.read_sql_query(sql,engine)

count
0     22

In [334]:
# Which stores have the highest sales of items over 90 proof? (edited) 

sql = """
SELECT vendor,SUM(item) as sum_item
FROM sales
GROUP BY vendor
ORDER BY 2 DESC


LIMIT 5
"""


pd.read_sql_query(sql,engine)

vendor     sum_item
0   Diageo Americas  19336983550
1    Luxco-St Louis  14242282206
2   Jim Beam Brands  12976648277
3  Sazerac Co. Inc.   8071285414
4           Proximo   7406338402

# A nice time for a Window Function maybe?

In [163]:
# Store 2238 (Adventureland Inn at 3200 Adventureland Dr) sold $883.24 in April and $27,526.38 in May, 
# for a 3017% growth rate.That was the highest percentage month-on-month growth rate. 
# Create a query that shows this and the next 9 highest after that


sql = """
SELECT vendor_no,vendor,(lead_total - total)/total * 100 as growth
FROM(SELECT vendor_no,vendor,month,total,COALESCE(LEAD(total) OVER (partition by vendor),0) as lead_total
FROM(SELECT vendor_no,vendor,DATE_PART('month',date) as month,SUM(total) as total
FROM sales
GROUP BY vendor_no,vendor,DATE_PART('month',date)
ORDER BY 2 ASC, 3 ASC)as temp)as temp_2
ORDER BY 3 DESC

LIMIT 15
"""
pd.read_sql_query(sql, engine)


vendor_no                             vendor       growth
0        207        Paradise Distilling Company  6413.814433
1        270   International Beverage Co., Inc.  4962.291412
2        381                 Windy Hill Spirits  4133.333333
3        207        Paradise Distilling Company  3965.476376
4        101           Artisan Grain Distillery  3285.714286
5         79            Breckenridge Distillery  3229.292929
6        111               Better Brands Bev Co  1520.000000
7        287                Chatham Imports,inc  1075.000000
8        962  Duggan's Distillers Products Corp  1044.396982
9        338                   Niche Import Co.  1033.781513
10       885          Yahara Bay Distillers Inc  1013.765326
11       417                Savvy Distillers LP   851.746242
12       497                WILDMAN AND SONS F.   754.586010
13       295                Kobrand Corporation   703.829698
14       226                   maDIKwe USA Inc.   687.201646

# Much more challenging.Pure Bonus level questions.

In [348]:
# We think the data might have been corrupted in some way. 
# The category listed in the sales table doesn’t always match up with the category in the products table.
# How many times has this happened, and can you find any patterns to it?


sql = """
WITH join_tab as (SELECT s.category_name as s_category, p.category_name as p_category
FROM sales as s
JOIN products as p
ON s.item = p.item_no)


SELECT SUM(not_equal_cat) as total_not_equal
FROM(SELECT s_category,p_category,COUNT(*)OVER () as not_equal_cat
FROM join_tab
WHERE s_category != p_category
GROUP BY s_category,p_category
ORDER BY 1 ASC) as temp
"""
pd.read_sql_query(sql, engine)



# The pattern that mismatch is maybe the person who key the sales order may select wrong.



total_not_equal
0            225.0

In [438]:
#  The store_address field in the stores table actually contains three rows of text. 
# Quite often the latitude and longitude are in the last line of text. 
# Create a query that shows the geo-locatable stores in latitude order 
# (i.e. show the stores from the most northerly to the most southerly).





sql = """
WITH SPLITED as (SELECT store,split_text ,ROW_NUMBER()OVER (partition by store) as row
FROM(SELECT store, unnest(string_to_array(store_address, '\n')) as split_text
FROM stores
GROUP BY store)as temp),

    SPLIT_COOR as (SELECT store,split_text ,row
    FROM SPLITED
    WHERE row = '3')

SELECT store,latitude,longtitude
FROM(SELECT * ,LEAD(latitude) OVER() as longtitude, ROW_NUMBER() OVER (partition by store)
FROM(SELECT store,unnest(string_to_array(TRIM(')' FROM TRIM('(' FROM split_text)) , ',')) as latitude
FROM SPLIT_COOR)as temp) as temp_2
WHERE row_number = 1
ORDER BY 2 DESC

LIMIT 10


"""


# STEP
# EXTRACT store_no, text
# SPLIT text into 3 row per store_no , also given row number
# SELECT row number 3 of each store_no
# SPLIT again the coordinate and lead the coordinate to match the lat,long in the same row, also give the row number 
# select only row number one which is correct order of lat, long for each store




pd.read_sql_query(sql,engine)

store            latitude           longtitude
0   3596   43.49114240400007   -92.36121936999996
1   3914   43.45480634900008   -95.32138065199996
2   4030   43.45480634900008   -95.32138065199996
3   4006   43.45384534900006   -96.43538427799996
4   4904   43.45355826500003   -96.43408946799997
5   4337   43.45331135500004   -96.42971810199998
6   4839   43.44498284600007   -93.22093823499995
7   3664  43.443225246000054   -93.21830107799997
8   3705   43.43186091100006   -96.17157696499999
9   3657   43.43104975500006   -96.16811881199999